NullPointerException when creating an index on an empty Map [SQL]

http://jenkins.hazelcast.com/view/stable/job/stable-x1/191/console

/disk1/workspace/stable-x1/5.1-SNAPSHOT/2021_12_13-10_09_09/sql-service

 CREATE INDEX mapBak1HD1ageIdx  ON mapBak1HD1 (age) TYPE SORTED
gives

com.hazelcast.sql.HazelcastSqlException
	at com.hazelcast.sql.impl.QueryUtils.toPublicException(QueryUtils.java:74)
	at com.hazelcast.sql.impl.SqlServiceImpl.execute(SqlServiceImpl.java:211)
	at com.hazelcast.sql.impl.SqlServiceImpl.execute(SqlServiceImpl.java:162)
	at com.hazelcast.sql.impl.SqlServiceImpl.execute(SqlServiceImpl.java:158)
	at com.hazelcast.sql.impl.SqlServiceImpl.execute(SqlServiceImpl.java:154)
	at com.hazelcast.sql.SqlService.execute(SqlService.java:89)
	at hzcmd.map.sql.service.CreateIndex.timeStep(CreateIndex.java:11)
at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.NullPointerException
	at com.hazelcast.jet.sql.impl.PlanExecutor.execute(PlanExecutor.java:133)
	at com.hazelcast.jet.sql.impl.SqlPlanImpl$CreateIndexPlan.execute(SqlPlanImpl.java:269)
	at com.hazelcast.sql.impl.SqlServiceImpl.query0(SqlServiceImpl.java:248)
	at com.hazelcast.sql.impl.SqlServiceImpl.execute(SqlServiceImpl.java:198)

--- a/hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java
+++ b/hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java
@@ -130,6 +130,12 @@ public class PlanExecutor {
             // common case. There's no atomic operation to create an index in IMDG, so it's not easy to
             // implement.
             MapContainer mapContainer = getMapContainer(hazelcastInstance.getMap(plan.mapName()));
+
+            // Very rare case, but still may happens...
+            if (mapContainer == null) {
+                throw QueryException.error("Can't create index: mapping '" + plan.mapName() + "' not initialized yet");
+            }
+
             if (mapContainer.getIndexes().getIndex(plan.indexName()) != null) {
                 throw QueryException.error("Can't create index: index '" + plan.indexName() + "' already exists");
             }

--- a/hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java
+++ b/hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java
@@ -131,11 +131,6 @@ public class PlanExecutor {
             // implement.
             MapContainer mapContainer = getMapContainer(hazelcastInstance.getMap(plan.mapName()));
 
-            // Very rare case, but still may happens...
-            if (mapContainer == null) {
-                throw QueryException.error("Can't create index: mapping '" + plan.mapName() + "' not initialized yet");
-            }
-
             if (mapContainer.getIndexes().getIndex(plan.indexName()) != null) {
                 throw QueryException.error("Can't create index: index '" + plan.indexName() + "' already exists");
             }
@@ -495,6 +490,6 @@ public class PlanExecutor {
         MapProxyImpl<K, V> mapProxy = (MapProxyImpl<K, V>) map;
         MapService mapService = mapProxy.getService();
         MapServiceContext mapServiceContext = mapService.getMapServiceContext();
-        return mapServiceContext.getMapContainers().get(map.getName());
+        return mapServiceContext.getMapContainer(map.getName());
     }
 }

--- a/hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlCreateIndexTest.java
+++ b/hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlCreateIndexTest.java
@@ -193,13 +193,17 @@ public class SqlCreateIndexTest extends OptimizerTestSupport {
     }
 
     private void checkPlan(boolean withIndex, boolean sorted, String sql) {
+        checkPlan(withIndex, sorted, sql, MAP_NAME);
+    }
+
+    private void checkPlan(boolean withIndex, boolean sorted, String sql, String mapName) {
         List<QueryDataType> parameterTypes = asList(QueryDataType.INT, QueryDataType.INT);
         List<TableField> mapTableFields = asList(
                 new MapTableField("__key", QueryDataType.INT, false, QueryPath.KEY_PATH),
                 new MapTableField("this", QueryDataType.INT, false, QueryPath.VALUE_PATH)
         );
         HazelcastTable table = partitionedTable(
-                MAP_NAME,
+                mapName,
                 mapTableFields,
                 getPartitionedMapIndexes(mapContainer(map), mapTableFields),
                 map.size()

--- a/hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlCreateIndexTest.java
+++ b/hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlCreateIndexTest.java
@@ -105,6 +105,17 @@ public class SqlCreateIndexTest extends OptimizerTestSupport {
     }
 
     @Test
+    public void when_indexCreatedWithEmptyMap_then_succeeds() {
+        String mapName = "map2";
+        String indexName = SqlTestSupport.randomName();
+        String sql = "CREATE INDEX IF NOT EXISTS " + indexName + " ON " + mapName + " (__key) TYPE SORTED";
+
+        instance().getSql().execute(sql);
+
+        assertThat(mapContainer(instance().getMap(mapName)).getIndexes().getIndex(indexName)).isNotNull();
+    }
+
+    @Test
     public void when_indexCreatedWithDefaultType_then_succeeds() {
         String indexName = SqlTestSupport.randomName();
         String selectSql = "SELECT * FROM " + MAP_NAME + " ORDER BY this DESC";

